RTech.CoffeeMachine.Api.Tests/BrewCoffee/Controllers/CoffeeMachineControllerTests.cs

```diff
@@ -120,29 +113,110 @@ public async Task GetBrewStatus_WhenCoffeeReady_ShouldReturn200()
+    {
+        var mockHttp = new MockHttpMessageHandler();
+        mockHttp.When("*")
+            .Throw(new HttpRequestException());
+
+        var httpClient = mockHttp.ToHttpClient();
+        httpClient.BaseAddress = new Uri(MOCKED_HTTPCLIENT_BASE_URL);
+        var response = await GetEndpointResponse(httpClient: httpClient);
+
+        response.StatusCode.ShouldBe(HttpStatusCode.BadGateway);
+        var content = await response.Content.ReadAsStringAsync();
+        content.ShouldBe(string.Empty);
+    }
+
+    [Fact]
+    public async Task GetBrewStatus_WhenTemperatureRequestReturns500_ShouldReturn502()
+    {
+        var mockHttp = new MockHttpMessageHandler();
+        mockHttp.When("*")
+            .Respond(HttpStatusCode.InternalServerError);
+
+        var httpClient = mockHttp.ToHttpClient();
+        httpClient.BaseAddress = new Uri(MOCKED_HTTPCLIENT_BASE_URL);
+        var response = await GetEndpointResponse(httpClient: httpClient);
+
+        response.StatusCode.ShouldBe(HttpStatusCode.BadGateway);
+        var content = await response.Content.ReadAsStringAsync();
+        content.ShouldBe(string.Empty);
+    }
+
+    [Theory]
+    [InlineData(30, BREW_READY_HOT_MESSAGE)]
+    [InlineData(30.01, "Your refreshing iced coffee is ready")]
+    public async Task GetBrewStatus_WhenTemperatureRetrieved_ShouldReturnCorrectMessage(float temperature, string expectedMessage)
+    {
+        var httpClient = CreateMockedHttpClient(temperature);
+        var response = await GetEndpointResponse(httpClient: httpClient);
+
+        response.StatusCode.ShouldBe(HttpStatusCode.OK);
+        var content = await response.Content.ReadFromJsonAsync<BrewStatusResponse>();
+        content.ShouldNotBeNull();
+        content.Message.ShouldBe(expectedMessage);
+    }
+
+    private Task<HttpResponseMessage> GetEndpointResponse(
+        IDateTimeProvider? dateTimeProvider = null,
+        IBrewStatusProvider? brewStatusProvider = null,
+        IMemoryCache? cache = null,
+        HttpClient? httpClient = null
+    )
     {
         var mockFactory = _factory.WithWebHostBuilder(builder =>
         {
-            builder.ConfigureTestServices(servicesConfiguration);
+            builder.ConfigureTestServices(services =>
+            {
+                if (dateTimeProvider != null)
+                    services.AddSingleton<IDateTimeProvider>(dateTimeProvider);
+                if (brewStatusProvider != null)
+                    services.AddSingleton<IBrewStatusProvider>(brewStatusProvider);
+                if (cache != null)
+                    services.AddSingleton<IMemoryCache>(cache);
+
+                httpClient ??= CreateMockedHttpClient(30);
+                services.AddHttpClient<IWeatherProxy, WeatherProxy>(client => new WeatherProxy(
+                    Mock.Of<ILogger<WeatherProxy>>(),
+                    httpClient
+                ));
+            });
```

Comment:
So the idea behind the `GetEndpointResponse` refactor was that it used to accept an `Action` parameter, which basically allowed integration test cases to handle the DI.

This PR now converts the `Action` parameter into a list of optional parameters. This means "heavy" DI container logic only appears in this 1 specific method in this file, & also ensuring that code duplication remains minimal.